gateway-js/src/supergraphManagers/UplinkSupergraphManager/index.ts

```diff
@@ -0,0 +1,225 @@
+    this.maxRetries = maxRetries ?? this.uplinkEndpoints.length * 3 - 1;
+
+    this.fetcher = fetcher ?? this.fetcher;
+    this.fallbackPollIntervalMs = fallbackPollIntervalInMs ?? this.fallbackPollIntervalMs;
+    this.shouldRunSubgraphHealthcheck =
+      shouldRunSubgraphHealthcheck ?? this.shouldRunSubgraphHealthcheck;
+    this.onFailureToFetchSupergraphSdl = onFailureToFetchSupergraphSdl?.bind(this);
+
+    this.state = { phase: 'initialized' };
+  }
+
+  public async initialize({ update, healthCheck }: SupergraphSdlHookOptions) {
+    this.update = update;
+
+    if (this.shouldRunSubgraphHealthcheck) {
+      this.healthCheck = healthCheck;
+    }
+
+    let initialSupergraphSdl: string | undefined = undefined;
+    try {
+      const result = await this.updateSupergraphSdl();
+      initialSupergraphSdl = result?.supergraphSdl;
+      if (result?.minDelaySeconds) {
+        this.minDelayMs = 1000 * result?.minDelaySeconds;
+        this.earliestFetchTime = new Date(Date.now() + this.minDelayMs);
+      }
+    } catch (e) {
+      this.logUpdateFailure(e);
+      throw e;
+    }
+
+    // Start polling after we resolve the first supergraph
+    this.beginPolling();
+
+    return {
+      // on init, this supergraphSdl should never actually be `undefined`.
+      // `this.updateSupergraphSdl()` will only return null if the schema hasn't
+      // changed over the course of an _update_.
+      supergraphSdl: initialSupergraphSdl ?? '',
+      cleanup: async () => {
+        if (this.state.phase === 'polling') {
+          await this.state.pollingPromise;
+        }
+        this.state = { phase: 'stopped' };
+        if (this.timerRef) {
+          clearTimeout(this.timerRef);
+          this.timerRef = null;
+        }
+      },
+    };
+  }
+
+  private async updateSupergraphSdl(): Promise<{
+    supergraphSdl: string;
+    minDelaySeconds?: number;
+  } | null> {
+    let supergraphSdl;
+    let minDelaySeconds: number | undefined = this.fallbackPollIntervalMs / 1000;
+
+    try {
+      const result = await loadSupergraphSdlFromUplinks({
+        graphRef: this.graphRef,
+        apiKey: this.apiKey,
+        endpoints: this.uplinkEndpoints,
+        errorReportingEndpoint: this.errorReportingEndpoint,
+        fetcher: this.fetcher,
+        compositionId: this.compositionId ?? null,
+        maxRetries: this.maxRetries,
+        roundRobinSeed: this.fetchCount++,
+        earliestFetchTime: this.earliestFetchTime,
+        logger: this.logger,
+      });
+
+      if (!result) {
+        return null;
+      }
+
+      this.compositionId = result.id;
+
+      ({ supergraphSdl, minDelaySeconds } = result);
+    } catch (e) {
+      if (!this.onFailureToFetchSupergraphSdl) {
+        throw e;
+      }
+
+      this.logger.debug(
+        'Error fetching supergraphSdl from Uplink, calling updateSupergraphSdlFailureCallback',
+      );
+      supergraphSdl = await this.onFailureToFetchSupergraphSdl({ error: e });
```

Comment:
Is there a way to tell if `onFailureToFetchSupergraphSdl` is called on startup, versus on a subsequent poll?

Most of the time, we want to provide a fallback supergraph _only on startup_. If Uplink is unreachable during a deploy or restart, the pod will crash. But it's fine if Uplink is unavailable for some period of time after the gateway is already running. 

I'm not sure I'd want to fallback to (possibly older) cached supergraph on momentary blips. I just want to make sure the pods can start up.

Reply:
Actually, now that I think about it — should there be a way for the `onFailureToFetchSupergraphSdl` to "no-op"? What happens if it doesn't return a string? I'd like a way to just let things poll after the gateway has started.

```js
onFailureToFetchSupergraphSdl({ error, hasStarted, graphRef }) {
  logger.warn(error);

  if (!hasStarted) {
    return fetchSupergraphFromS3(graphRef);
  }

  return null; // let the gateway continue polling until Uplink is available again
}
```

Or maybe the type of error could help indicate the state of the gateway?

```js
switch (error.type) {
  case 'UplinkPollFailed':
  case 'GatewayStartedFailed':
}
```

Reply:
lol jinx

Reply:
Awesome feedback, thank you!

I added callbacks for `onFailureToFetchSupergraphSdlDuringInit` and `onFailureToFetchSupergraphSdlAfterInit`, with the biggest difference being the latter can `return null` to signify to the gateway to reuse the existing schema (other than being called at different times).

I made these the params to the callbacks, but happy to expose other info that's needed/makes sense

```ts
export type FailureToFetchSupergraphSdlFunctionParams = {
  error: Error;
  graphRef: string;
  logger: Logger;
  fetchCount: number;
};
```

Reply:
sounds great! 👍🏻